### PR TITLE
Support templated LLVM intrinsics with vector arguments

### LINKED ISSUE #2962: ICE: intrinsic with vector arguments
Testcase that segfaults:
```d
import core.simd;

pragma(LDC_intrinsic, "llvm.fshl.i#")
    T llvm_fshl(T)(T a, T b, T shift)
        if (__traits(isIntegral, T));

int4 rotateLeftIntrinsic(int4 x, int4 bits) {
    return llvm_fshl(x, x, bits);
}
```
https://d.godbolt.org/z/XUvSX9

Apparently vector types pass the `isIntegral` test. At the least, we shouldn't crash on this.

### PR DESCRIPTION
**Pull request by kinke**
Fixes #2962.

### CODE REVIEW

File: gen/tollvm.cpp [hunk 1]
```diff
@@ -187,17 +187,11 @@ LLType *DtoMemType(Type *t) { return i1ToI8(voidToI8(DtoType(t))); }
 LLPointerType *DtoPtrToType(Type *t) { return DtoMemType(t)->getPointerTo(); }
 
 LLType *voidToI8(LLType *t) {
-  if (t == LLType::getVoidTy(gIR->context())) {
-    return LLType::getInt8Ty(gIR->context());
-  }
-  return t;
+  return t->isVoidTy() ? LLType::getInt8Ty(t->getContext()) : t;
```
Comment (kinke) [inline gen/tollvm.cpp]: `gIR` wasn't set yet, hence the crash.

File: gen/llvmhelpers.cpp [hunk 2]
```diff
@@ -1309,25 +1309,27 @@ static char *DtoOverloadedIntrinsicName(TemplateInstance *ti,
...
-          static_cast<unsigned long>(gDataLayout->getTypeSizeInBits(dtype)));
-
-  // replace # in name with bitsize
-  std::string name(td->intrinsicName);
+  std::string name = td->intrinsicName;
+
+  // replace "{f,i}#" by "{f,i}<bitsize>" or "v<vector length>{f,i}<vector
+  // element bitsize>"
+  llvm::Type *dtype = DtoType(T);
+  char replacement[24]; // probably excessive, but covers a uint64_t
```
Comment (JohanEngelen) [inline gen/llvmhelpers.cpp]: I don't understand the comment.

Safer to use `llvm::SmallString` and `llvm::formatv` for the replacement string.

Comment (JohanEngelen) [reply gen/llvmhelpers.cpp]: besides this, lgtm

Comment (kinke) [reply gen/llvmhelpers.cpp]: What isn't understandable, the `{f,i}` bash-style set?

Comment (kinke) [reply gen/llvmhelpers.cpp]: Argh, `llvm::formatv()` not available for 3.9.

**Comment by JohanEngelen** [comment]
The tests needs to do linking aswell, such that it would fail if we'd get the intrinsic name wrong as mentioned here: https://github.com/ldc-developers/druntime/pull/151#issuecomment-455828670

**Comment by kinke** [comment]
[LLVM 7+ errors out if the name doesn't match the signature, that's how I got aware of that missing piece.]

I'll look for another intrinsic and link.

**Comment by kinke** [comment]
Slightly reworked, incl. fix for FP vectors (`TypeVector` doesn't forward `isreal()` to its element type, `isfloating()` does).

File: gen/llvmhelpers.cpp [hunk 3]
```diff
@@ -1302,32 +1302,39 @@ static char *DtoOverloadedIntrinsicName(TemplateInstance *ti,
...
+    llvm::raw_string_ostream stream(replacement);
+    stream << 'v' << dtype->getVectorNumElements() << prefix
+        << gDataLayout->getTypeSizeInBits(dtype->getVectorElementType());
+    stream.flush();
+  } else {
+    replacement = prefix + std::to_string(gDataLayout->getTypeSizeInBits(dtype));
+  }
 
-  std::string needle("#");
+  const std::string needle = "#";
```
Comment (Hardcode84) [inline gen/llvmhelpers.cpp]: string::find support const char* overload I think

**Pull request merged**